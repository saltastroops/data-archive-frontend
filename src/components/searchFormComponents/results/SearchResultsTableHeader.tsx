import { faSortDown, faSortUp } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import * as React from "react";
import { SortDirection, SortDirectionType } from "react-virtualized";

interface ISearchResultsTableHeaderProps {
  dataKey: string;
  sort: ({
    sortBy,
    sortDirection
  }: {
    sortBy: string;
    sortDirection: SortDirectionType;
  }) => void;
  sortBy?: string;
  style: object;
}

interface ISearchResultsTableHeaderState {
  nextSortDirection: SortDirectionType;
  previousSortBy?: string;
}

class SearchResultsTableHeader extends React.Component<
  ISearchResultsTableHeaderProps,
  ISearchResultsTableHeaderState
> {
  static getDerivedStateFromProps(
    props: ISearchResultsTableHeaderProps,
    state: ISearchResultsTableHeaderState
  ) {
    if (props.sortBy !== state.previousSortBy) {
      return {
        nextSortDirection:
          props.sortBy === props.dataKey
            ? SortDirection.DESC
            : SortDirection.ASC,
        previousSortBy: props.sortBy
      };
    } else {
      return null;
    }
  }
  constructor(props: ISearchResultsTableHeaderProps) {
    super(props);

    this.state = { nextSortDirection: SortDirection.ASC };
  }

  render() {
    const { children, dataKey, sortBy, style } = this.props;

    const showSortIndicator = sortBy === dataKey;

    return (
      <div
        className="search-results cell header"
        data-test="header"
        style={style}
        onClick={this.onClick}
      >
        {children}
        {showSortIndicator && (
          <FontAwesomeIcon
            icon={
              this.state.nextSortDirection === SortDirection.DESC
                ? faSortUp
                : faSortDown
            }
            style={{ paddingLeft: 3 }}
          />
        )}
      </div>
    );
  }

  onClick = () => {
    const nextSortDirection = this.state.nextSortDirection || SortDirection.ASC;
    const currentSortDirection =
      nextSortDirection === SortDirection.ASC
        ? SortDirection.DESC
        : SortDirection.ASC;

    this.props.sort({
      sortBy: this.props.dataKey,
      sortDirection: nextSortDirection
    });

    this.setState(() => ({ nextSortDirection: currentSortDirection }));
  };
}

export default SearchResultsTableHeader;
