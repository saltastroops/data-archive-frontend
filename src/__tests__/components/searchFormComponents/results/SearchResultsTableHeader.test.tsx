import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import { SortDirection } from "react-virtualized";
import wait from "waait";
import SearchResultsTableHeader from "../../../../components/searchFormComponents/results/SearchResultsTableHeader";

describe("SearchResultsTableHeader", () => {
  it("should render correctly", () => {
    const wrapper = mount(
      <SearchResultsTableHeader
        dataKey="ProposalTitle"
        sort={jest.fn()}
        style={{}}
      >
        <span>Proposal Title</span>
      </SearchResultsTableHeader>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render a sort direction indicator if the data key and sort by key are equal", () => {
    const wrapper = mount(
      <SearchResultsTableHeader
        dataKey="ProposalTitle"
        sort={jest.fn()}
        sortBy="ProposalTitle"
        style={{}}
      >
        <span>Proposal Title</span>
      </SearchResultsTableHeader>
    );
    expect(wrapper.find("svg").length).toBe(1);
  });

  it("should not render a sort direction indicator if the data key and sort by key are not equal", () => {
    const wrapper = mount(
      <SearchResultsTableHeader
        dataKey="ProposalTitle"
        sort={jest.fn()}
        style={{}}
      >
        <span>Proposal Title</span>
      </SearchResultsTableHeader>
    );
    expect(wrapper.find("svg").length).toBe(0);
  });

  it("should handle clicks on the header if the data key and sort by key are equal", async () => {
    // Initially there is an up indicator.
    const sort = jest.fn();
    const wrapper = mount(
      <SearchResultsTableHeader
        dataKey="ProposalTitle"
        sort={sort}
        sortBy="ProposalTitle"
        style={{}}
      >
        <span>Proposal Title</span>
      </SearchResultsTableHeader>
    );
    expect(wrapper.find("svg.fa-sort-up").length).toBe(1);

    // Click the header
    const header = wrapper.find('[data-test="header"]');
    header.simulate("click");
    await wait(0);
    wrapper.update();

    // There is a down indicator now
    expect(wrapper.find("svg.fa-sort-down").length).toBe(1);

    // Click the header again
    header.simulate("click");
    await wait(0);
    wrapper.update();

    // The sort function has been called twice
    expect(sort).toHaveBeenCalledTimes(2);
    expect(sort).toHaveBeenNthCalledWith(1, {
      sortBy: "ProposalTitle",
      sortDirection: SortDirection.DESC
    });
    expect(sort).toHaveBeenNthCalledWith(2, {
      sortBy: "ProposalTitle",
      sortDirection: SortDirection.ASC
    });
  });

  it("should handle clicks on the header if the data key and sort by key are not equal", async () => {
    const sort = jest.fn();
    const wrapper = mount(
      <SearchResultsTableHeader
        dataKey="ProposalTitle"
        sort={sort}
        sortBy="TargetName"
        style={{}}
      >
        <span>Proposal Title</span>
      </SearchResultsTableHeader>
    );

    // Click the header
    const header = wrapper.find('[data-test="header"]');
    header.simulate("click");
    await wait(0);
    wrapper.update();

    // The sort function has been called
    expect(sort).toHaveBeenCalledWith({
      sortBy: "ProposalTitle",
      sortDirection: SortDirection.ASC
    });
  });
});
