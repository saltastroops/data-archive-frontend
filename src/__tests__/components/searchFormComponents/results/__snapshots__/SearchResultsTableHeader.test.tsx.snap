// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsTableHeader should render correctly 1`] = `
<SearchResultsTableHeader
  dataKey="ProposalTitle"
  sort={[MockFunction]}
  style={Object {}}
>
  <div
    className="search-results cell header"
    data-test="header"
    onClick={[Function]}
    style={Object {}}
  >
    <span>
      Proposal Title
    </span>
  </div>
</SearchResultsTableHeader>
`;
